### ISSUE TITLE
'az login' hangs in linux server

### ISSUE THREAD

**Issue by violascarlett** [Azure/azure-cli#26109]
**Related command**

**Describe the bug**

**Expected behavior**

**Environment summary**

**Additional context**

**Comment by yonzhan** [comment]
Thank you for opening this issue, we will look into it.

**Comment by jiasli** [comment]
I can't repro this issue with a clean CentOS 7 container:

```
docker run -it --rm centos:7

echo -e "[azure-cli]
name=Azure CLI
baseurl=https://packages.microsoft.com/yumrepos/azure-cli
enabled=1
gpgcheck=1
gpgkey=https://packages.microsoft.com/keys/microsoft.asc" | tee /etc/yum.repos.d/azure-cli.repo

yum install azure-cli

az login
```

In your output, it looks like `msal_extensions` reached these lines in `CrossPlatLock._try_to_create_lock_file`: https://github.com/AzureAD/microsoft-authentication-extensions-for-python/blob/69eef72b0e7a540ae00abb6a65e749cfce4d532a/msal_extensions/cache_lock.py#L49-L51

`CrossPlatLock._try_to_create_lock_file` is called by `CrossPlatLock.__enter__`: https://github.com/AzureAD/microsoft-authentication-extensions-for-python/blob/69eef72b0e7a540ae00abb6a65e749cfce4d532a/msal_extensions/cache_lock.py#L57

`CrossPlatLock.__enter__` is only called once by `PersistedTokenCache.modify`: https://github.com/AzureAD/microsoft-authentication-extensions-for-python/blob/d0696aeb6f65168b1e0d405cd871b80bb101cd76/msal_extensions/token_cache.py#L63


Could you try to clear `~/.azure` (`/home/username/.azure`), instead of `/home/.azure`?

If that doesn't work, any chance you are running multiple Azure CLI instances concurrently?

cc @rayluo

**Comment by violascarlett** [comment]
@jiasli 
Tried cleaning /home/(username)/.azure directory but no changes.

could you be more specific about azure CLI instances? Do you mean containers?

FYI We are running azure CLI commands from different user accounts concurrently on a single server.

**Comment by jiasli** [comment]
> could you be more specific about azure CLI instances? Do you mean containers?

I mean [process](https://en.wikipedia.org/wiki/Process_(computing)).

Do different users share the same `/home/(username)/.azure` folder?

**Comment by rayluo** [comment]
> This happens executing from user accounts except root user.

Could it be you accidentally (or even deliberately) `sudo` once so that some helper file was created with root permission, therefore you non-root process won't be able to access it?

I know you said you already clean up the `/home/(username)/.azure`, but if your setup has one az instance running as root, the situation described above would re-occur.

**Comment by violascarlett** [comment]
> > could you be more specific about azure CLI instances? Do you mean containers?
> 
> I mean [process](https://en.wikipedia.org/wiki/Process_(computing)).
> 
> Do different users share the same `/home/(username)/.azure` folder?

No. Every users have different directories inside /home. And their only .azure directory stored.

I can figure out multiple processes are using azure cli.
<img width="642" alt="IMG_2615" src="https://user-images.githubusercontent.com/52439238/231751182-0179ede9-d376-414b-97aa-bfaed1cd8f9d.png">

**Comment by violascarlett** [comment]
> > This happens executing from user accounts except root user.
> 
> Could it be you accidentally (or even deliberately) `sudo` once so that some helper file was created with root permission, therefore you non-root process won't be able to access it?
> 
> I know you said you already clean up the `/home/(username)/.azure`, but if your setup has one az instance running as root, the situation described above would re-occur.

Most probably it may happened. Maybe done from other admins. So how can I fix this gracefully?

**Comment by rayluo** [comment]
> > Could it be you accidentally (or even deliberately) `sudo` once so that some helper file was created with root permission, therefore you non-root process won't be able to access it?
>
> Most probably it may happened. Maybe done from other admins. So how can I fix this gracefully?

Let's validate this assumption first. When the issue occurs, before you delete `/home/(username)/.azure`, what are the files owners and permissions in that folder? If they were owned by root, why did your another admin need "sudo az login ..."?

> > Do different users share the same `/home/(username)/.azure` folder?
> 
> No. Every users have different directories inside /home. And their only .azure directory stored.
> 
> I can figure out multiple processes are using azure cli. <img alt="IMG_2615" width="642" src="https://user-images.githubusercontent.com/52439238/231751182-0179ede9-d376-414b-97aa-bfaed1cd8f9d.png">
> 
> The process falling to uninterrupted sleep state (D).

"az login ..." is meant to be interactive. What happened to those now-sleeping processes? Did it go unresponsive so that you tried another one in a different terminal, which could then run into racing conditions with the first one? We better triage the first unresponsive "az login --debug" first. Kill the existing processes and start afresh, report back when with the "az login --debug" of your first failing login attempt.

**Comment by violascarlett** [comment]
> > > Could it be you accidentally (or even deliberately) `sudo` once so that some helper file was created with root permission, therefore you non-root process won't be able to access it?
> > 
> > 
> > Most probably it may happened. Maybe done from other admins. So how can I fix this gracefully?
> 
> Let's validate this assumption first. When the issue occurs, before you delete `/home/(username)/.azure`, what are the files owners and permissions in that folder? If they were owned by root, why did your another admin need "sudo az login ..."?
> 
The files permissions and ownership is belongs to the particular user only. I am not sure if ‘sudo az login’ used or not. Most probably it wouldn’t.

> > > Do different users share the same `/home/(username)/.azure` folder?
> > 
> > 
> > No. Every users have different directories inside /home. And their only .azure directory stored.
> > I can figure out multiple processes are using azure cli. <img alt="IMG_2615" width="642" src="https://user-images.githubusercontent.com/52439238/231751182-0179ede9-d376-414b-97aa-bfaed1cd8f9d.png">
> > The process falling to uninterrupted sleep state (D).
> 
> "az login ..." is meant to be interactive. What happened to those now-sleeping processes? Did it go unresponsive so that you tried another one in a different terminal, which could then run into racing conditions with the first one? We better triage the first unresponsive "az login --debug" first. Kill the existing processes and start afresh, report back when with the "az login --debug" of your first failing login attempt.
 
 Exactly, I tried multiple times. Not from different terminal. After a process stuck I use ctrl+z to stop it and did again.

 The process state D cannot be killed. Server restart can vanish these processes. I did that. And again the ‘az login —debug’ also going to uninterruptible sleep. 
 
And the best part is I did reinstalling Azure cli by ‘yum reinstall azure-cli’

**Comment by rayluo** [comment]
> > "az login ..." is meant to be interactive. What happened to those now-sleeping processes? Did it go unresponsive so that you tried another one in a different terminal, which could then run into racing conditions with the first one? We better triage the first unresponsive "az login --debug" first. Kill the existing processes and start afresh, report back when with the "az login --debug" of your first failing login attempt.
> 
> Exactly, I tried multiple times. Not from different terminal. After a process stuck I use ctrl+z to stop it and did again.
> 
> The process state D cannot be killed. Server restart can vanish these processes. I did that. And again the ‘az login —debug’ also going to uninterruptible sleep.

Did you mean, after you killed all existing processes and started afresh, your very first "az login --debug" will hang? And it is reproducible every time?

Ideally, we want the debug logs of your first hanging process. So, you better stick with the `--debug` during your investigation.

**Comment by violascarlett** [comment]
> > > "az login ..." is meant to be interactive. What happened to those now-sleeping processes? Did it go unresponsive so that you tried another one in a different terminal, which could then run into racing conditions with the first one? We better triage the first unresponsive "az login --debug" first. Kill the existing processes and start afresh, report back when with the "az login --debug" of your first failing login attempt.
> > 
> > 
> > Exactly, I tried multiple times. Not from different terminal. After a process stuck I use ctrl+z to stop it and did again.
> > The process state D cannot be killed. Server restart can vanish these processes. I did that. And again the ‘az login —debug’ also going to uninterruptible sleep.
> 
> Did you mean, after you killed all existing processes and started afresh, your very first "az login --debug" will hang? And it is reproducible every time?
> 
Yes exactly.
> Ideally, we want the debug logs of your first hanging process. So, you better stick with the `--debug` during your investigation.

$ az login --use-device-code --debug
cli.knack.cli: Command arguments: ['login', '--use-device-code', '--debug']
cli.knack.cli: __init__ debug log:
Enable color in terminal.
cli.knack.cli: Event: Cli.PreExecute []
cli.knack.cli: Event: CommandParser.OnGlobalArgumentsCreate [<function CLILogging.on_global_arguments at 0x7f78343a8048>, <function OutputProducer.on_global_arguments at 0x7f7833f07ea0>, <function CLIQuery.on_global_arguments at 0x7f7833c98400>]
cli.knack.cli: Event: CommandInvoker.OnPreCommandTableCreate []
cli.azure.cli.core: Modules found from index for 'login': ['azure.cli.command_modules.profile']
cli.azure.cli.core: Loading command modules:
cli.azure.cli.core: Name                  Load Time    Groups  Commands
cli.azure.cli.core: profile                   0.002         2         9
cli.azure.cli.core: Total (1)                 0.002         2         9
cli.azure.cli.core: Loaded 2 groups, 9 commands.
cli.azure.cli.core: Found a match in the command table.
cli.azure.cli.core: Raw command  : login
cli.azure.cli.core: Command table: login
cli.knack.cli: Event: CommandInvoker.OnPreCommandTableTruncate [<function AzCliLogging.init_command_file_logging at 0x7f7832c4a598>]
cli.azure.cli.core.azlogging: metadata file logging enabled - writing logs to '/home/localad.tech/<username>/.azure/commands/2023-04-18.15-45-04.login.672821.log'.
az_command_data_logger: command args: login --use-device-code --debug
cli.knack.cli: Event: CommandInvoker.OnPreArgumentLoad [<function register_global_subscription_argument.<locals>.add_subscription_parameter at 0x7f78327d3268>]
cli.knack.cli: Event: CommandInvoker.OnPostArgumentLoad []
cli.knack.cli: Event: CommandInvoker.OnPostCommandTableCreate [<function register_ids_argument.<locals>.add_ids_arguments at 0x7f7832724840>, <function register_cache_arguments.<locals>.add_cache_arguments at 0x7f7832724950>]
cli.knack.cli: Event: CommandInvoker.OnCommandTableLoaded []
cli.knack.cli: Event: CommandInvoker.OnPreParseArgs []
cli.knack.cli: Event: CommandInvoker.OnPostParseArgs [<function OutputProducer.handle_output_argument at 0x7f7833f07f28>, <function CLIQuery.handle_query_parameter at 0x7f7833c98488>, <function register_ids_argument.<locals>.parse_ids_arguments at 0x7f78327248c8>]
cli.azure.cli.core.auth.persistence: build_persistence: location='/home/localad.tech/<username>/.azure/msal_token_cache.json', encrypt=False
cli.azure.cli.core.auth.binary_cache: load: /home/localad.tech/skrishnan010/.azure/msal_http_cache.bin
urllib3.util.retry: Converted retries value: 1 -> Retry(total=1, connect=None, read=None, redirect=None, status=None)
cli.azure.cli.core.auth.binary_cache: save: /home/localad.tech/<username>/.azure/msal_http_cache.bin
cli.azure.cli.core.auth.binary_cache: save: /home/localad.tech/<username>/.azure/msal_http_cache.bin
urllib3.connectionpool: Starting new HTTPS connection (1): login.microsoftonline.com:443
urllib3.connectionpool: https://login.microsoftonline.com:443 "GET /organizations/v2.0/.well-known/openid-configuration HTTP/1.1" 200 1589
cli.azure.cli.core.auth.binary_cache: save: /home/localad.tech/<username>/.azure/msal_http_cache.bin
cli.azure.cli.core.auth.binary_cache: save: /home/localad.tech/<username>/.azure/msal_http_cache.bin
msal.authority: openid_config = {'token_endpoint': 'https://login.microsoftonline.com/organizations/oauth2/v2.0/token', 'token_endpoint_auth_methods_supported': ['client_secret_post', 'private_key_jwt', 'client_secret_basic'], 'jwks_uri': 'https://login.microsoftonline.com/organizations/discovery/v2.0/keys', 'response_modes_supported': ['query', 'fragment', 'form_post'], 'subject_types_supported': ['pairwise'], 'id_token_signing_alg_values_supported': ['RS256'], 'response_types_supported': ['code', 'id_token', 'code id_token', 'id_token token'], 'scopes_supported': ['openid', 'profile', 'email', 'offline_access'], 'issuer': 'https://login.microsoftonline.com/{tenantid}/v2.0', 'request_uri_parameter_supported': False, 'userinfo_endpoint': 'https://graph.microsoft.com/oidc/userinfo', 'authorization_endpoint': 'https://login.microsoftonline.com/organizations/oauth2/v2.0/authorize', 'device_authorization_endpoint': 'https://login.microsoftonline.com/organizations/oauth2/v2.0/devicecode', 'http_logout_supported': True, 'frontchannel_logout_supported': True, 'end_session_endpoint': 'https://login.microsoftonline.com/organizations/oauth2/v2.0/logout', 'claims_supported': ['sub', 'iss', 'cloud_instance_name', 'cloud_instance_host_name', 'cloud_graph_host_name', 'msgraph_host', 'aud', 'exp', 'iat', 'auth_time', 'acr', 'nonce', 'preferred_username', 'name', 'tid', 'ver', 'at_hash', 'c_hash', 'email'], 'kerberos_endpoint': 'https://login.microsoftonline.com/organizations/kerberos', 'tenant_region_scope': None, 'cloud_instance_name': 'microsoftonline.com', 'cloud_graph_host_name': 'graph.windows.net', 'msgraph_host': 'graph.microsoft.com', 'rbac_url': 'https://pas.windows.net'}
urllib3.connectionpool: https://login.microsoftonline.com:443 "POST /organizations/oauth2/v2.0/devicecode HTTP/1.1" 200 473
cli.azure.cli.core.auth.identity: To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code FC23YSKMB to authenticate.
msal.telemetry: Generate or reuse correlation_id: 801eec5e-d0fb-4da7-9f4c-dd9a15f8870a
urllib3.connectionpool: https://login.microsoftonline.com:443 "POST /organizations/oauth2/v2.0/token HTTP/1.1" 400 510
FC23YSKMBurllib3.connectionpool: https://login.microsoftonline.com:443 "POST /organizations/oauth2/v2.0/token HTTP/1.1" 400 510                                                                                                          urllib3.connectionpool: https://login.microsoftonline.com:443 "POST /organizations/oauth2/v2.0/token HTTP/1.1" 400 510
urllib3.connectionpool: https://login.microsoftonline.com:443 "POST /organizations/oauth2/v2.0/token HTTP/1.1" 400 510
urllib3.connectionpool: https://login.microsoftonline.com:443 "POST /organizations/oauth2/v2.0/token HTTP/1.1" 400 510
urllib3.connectionpool: https://login.microsoftonline.com:443 "POST /organizations/oauth2/v2.0/token HTTP/1.1" 400 510
urllib3.connectionpool: https://login.microsoftonline.com:443 "POST /organizations/oauth2/v2.0/token HTTP/1.1" 400 510
urllib3.connectionpool: https://login.microsoftonline.com:443 "POST /organizations/oauth2/v2.0/token HTTP/1.1" 400 510
urllib3.connectionpool: https://login.microsoftonline.com:443 "POST /organizations/oauth2/v2.0/token HTTP/1.1" 200 5559
msal_extensions.cache_lock: Process 672821 found existing lock file, will retry after 0.250000 second
msal_extensions.cache_lock: Process 672821 found existing lock file, will retry after 0.250000 second
msal_extensions.cache_lock: Process 672821 found existing lock file, will retry after 0.250000 second
msal_extensions.cache_lock: Process 672821 found existing lock file, will retry after 0.250000 second
msal_extensions.cache_lock: Process 672821 found existing lock file, will retry after 0.250000 second
msal_extensions.cache_lock: Process 672821 found existing lock file, will retry after 0.250000 second
msal_extensions.cache_lock: Process 672821 found existing lock file, will retry after 0.250000 second
msal_extensions.cache_lock: Process 672821 found existing lock file, will retry after 0.250000 second
msal_extensions.cache_lock: Process 672821 found existing lock file, will retry after 0.250000 second
msal_extensions.cache_lock: Process 672821 found existing lock file, will retry after 0.250000 second
msal_extensions.cache_lock: Process 672821 found existing lock file, will retry after 0.250000 second
msal_extensions.cache_lock: Process 672821 found existing lock file, will retry after 0.250000 second
msal_extensions.cache_lock: Process 672821 found existing lock file, will retry after 0.250000 second
msal_extensions.cache_lock: Process 672821 found existing lock file, will retry after 0.250000 second
msal_extensions.cache_lock: Process 672821 found existing lock file, will retry after 0.250000 second
msal_extensions.cache_lock: Process 672821 found existing lock file, will retry after 0.250000 second
msal_extensions.cache_lock: Process 672821 found existing lock file, will retry after 0.250000 second
msal_extensions.cache_lock: Process 672821 found existing lock file, will retry after 0.250000 second
msal_extensions.cache_lock: Process 672821 found existing lock file, will retry after 0.250000 second
msal_extensions.cache_lock: Process 672821 found existing lock file, will retry after 0.250000 second
msal_extensions.cache_lock: Process 672821 failed to create lock file


<img width="725" alt="IMG_2662" src="https://user-images.githubusercontent.com/52439238/232874361-f9202382-4fad-4d2a-9ae3-12e791c4622c.png">

**Comment by rayluo** [comment]
> > **Environment summary**
> > CLI version (`az --version : 2.38.0`) / OS : RHEL 7 / Shell Type : bash
>
> I can't repro this issue with a clean CentOS 7 container:
> 
> ```
> docker run -it --rm centos:7
> ```

How can we reproduce this? @violascarlett, are you using a certain container image, or a specific SKU of Azure VM etc.?

**Comment by violascarlett** [comment]
@rayluo yeah we are using Azure VM, installed RHEL 7.9 OS. And I am not sure how to reproduce this issue🙁. 

If nothing helps can I try removing Azure cli and install it again? But #yum reinstall azure-cli didn’t help.

**Comment by violascarlett** [comment]
@rayluo I tried installing azure cli version 2.29.0 and it is working. I think something wrong with the msal authentication.
